File: python_playground/search/uniform_cost_search_UCS.py
# Perform Uniform Cost Search (UCS) on a weighted graph (adjacency list) to find
# the least-cost path from a start node to a goal node. UCS uses a priority queue
# to always expand the node with the lowest cumulative cost first.

# Assumes the graph is positively weighted (no negative edge weights).
# Returns (path, cost) if the goal is reachable; otherwise (None, -1).

# Time Complexity: O(b^d)
# - b: branching factor
# - d: depth of the optimal solution (in terms of number of edges, not cost)

# Space Complexity: O(b^d)
# - Due to storing all frontier paths in the priority queue

# Example:
# graph = {
#     'A': [('B', 1), ('C', 5)],
#     'B': [('D', 2)],
#     'C': [('D', 2)],
#     'D': []
# }
# uniform_cost_search(graph, 'A', 'D') → (['A', 'B', 'D'], 3)

import heapq

class UCS:
    def uniform_cost_search(self, graph, start, goal):
        queue = [(0, start, [start])]    # Priority queue: (cost_so_far, current_node, path_to_node)
        visited = set()
        while queue:
            # The heapq module implements a min-heap, which keeps the smallest element always at the top
            cost, node, path = heapq.heappop(queue)    # Pop the smallest element O(log n)
            if node in visited:
                continue
            visited.add(node)
            if node == goal:
                return path, cost
            for neighbor, weight in graph.get(node, []):
                if neighbor not in visited:
                    heapq.heappush(queue, (cost + weight, neighbor, path + [neighbor])) # Add a new node O(log n)
        return None, -1

File: cpp_playground/search/uniform_cost_search_UCS.cpp
/**
 Perform Uniform Cost Search (UCS) on a weighted graph (adjacency list) to find
 the least-cost path from a start node to a goal node.

 UCS uses a priority queue to always expand the node with the lowest cumulative cost first.
 It guarantees an optimal path when all edge weights are non-negative.

 Assumptions:
 - The graph is positively weighted (no negative edge weights).
 
 Returns:
 - A pair (path, cost) where `path` is a vector of node names (std::string) and `cost` is the total path cost.
 - If the goal is unreachable, returns (empty vector, -1).
 
 Time Complexity: O(b^d)
 - b: branching factor
 - d: depth of the optimal solution (in terms of number of edges)
 
 Space Complexity: O(b^d)
 - Due to storing all frontier paths in the priority queue
 
 Example:
  UCS::Graph graph = {
     {"A", {{"B", 1}, {"C", 5}}},
     {"B", {{"D", 2}}},
     {"C", {{"D", 2}}},
    {"D", {}}
  };
  UCS ucs;
  auto [path, cost] = ucs.uniform_cost_search(graph, "A", "D");
  result --> path = {"A", "B", "D"}, cost = 3
 */














#include "uniform_cost_search_UCS.hpp"

UCS::PathResult UCS::uniform_cost_search(const Graph& graph, const Node& start, const Node& goal){
    using QueueNode = std::tuple<int, Node, std::vector<Node>>;   // (cost_so_far, current_node_name, path_to_node)
    // Lambda comparator: sort by cost (lowest first)
    auto comparator = [](const QueueNode& a, const QueueNode& b){
        return std::get<0>(a) > std::get<0>(b);   // min-heap: smaller cost has higher priority
    };
    // std::priority_queue is by default a max-heap, meaning it always pops the largest element first, so comparator function is applied.
    std::priority_queue<QueueNode, std::vector<QueueNode>, decltype(comparator)> queue(comparator);
    std::set<Node> visited;

    queue.emplace(0, start, std::vector<Node>{start});

    while (!queue.empty()){
        auto [cost, node, path] = queue.top();
        queue.pop();
        if (visited.contains(node)) continue;
        visited.insert(node);
        if (node == goal) return {path, cost};
        if (graph.contains(node)) {
            for (const auto& [neighbor, weight] : graph.at(node)) {
                if (!visited.contains(neighbor)) {
                    auto new_path = path;
                    new_path.push_back(neighbor);
                    queue.emplace(cost + weight, neighbor, new_path);
                }
            }
        }
    }
    return {{}, -1}; 
}